File: ABC/ABC112/D/main.cpp
#include <bits/stdc++.h>
using namespace std;

vector<long> divisor(long N) {
  vector<long> res;
  for (long i = 1; i * i <= N; i++) {
    if (N % i == 0) {
      res.push_back(i);
      if (i != N / i) {
        res.push_back(N / i);
      }
    }
  }
  return res;
}

int main() {

  int N, M;
  cin >> N >> M;

  int d = M / N;
  
  vector<long> v = divisor(M);

  long answer = 0;

  for (int i = 0; i < v.size(); i++) {
    if (v[i] <= d) {
      answer = max(answer, v[i]);
    }
  }

  cout << answer << endl;

}

File: ABC/ABC112/D/main.java
import java.util.*;

public class Main {

  public static void main (String[] args) {

    Scanner sc = new Scanner(System.in);
    long N = sc.nextLong();
    long M = sc.nextLong();
    sc.close();
  
    long d = M / N;

    Set<Long> v = divisor(M);

    long answer = 0;
 
    for (Long value : v) {
      if (value <= d) {
        answer = Math.max(answer, value);
      }
    }

    System.out.println(answer);

  }

  static Set<Long> divisor (long N) {
    Set<Long> res = new HashSet<Long>();
    for (long i = 1; i * i <= N; i++) {
      if (N % i == 0) {
        res.add(i);
        if (i != N / i) {
          res.add(N / i);
        }
      }
    }
    return res;
  }

}